File: leetcode/lc-us/cpp/1807.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    string evaluate(string s, vector<vector<string>> &knowledge) {
        unordered_map<string, string> map;
        for (auto &k: knowledge) {
            map[k[0]] = k[1];
        }
        string res;
        int n = (int) s.size();
        for (int i = 0; i < n;) {
            if (s[i] != '(') {
                res.push_back(s[i]);
                i++;
            } else {
                int j = i;
                while (j < n && s[j] != ')') {
                    j++;
                }
                string key = s.substr(i + 1, j - i - 1);
                if (map.find(key) != map.end()) {
                    res += map[key];
                } else {
                    res.push_back('?');
                }
                i = j + 1;
            }
        }
        return res;
    }
};

File: leetcode/lc-us/py/1807.py
from typing import List


class Solution:
    def evaluate(self, s: str, knowledge: List[List[str]]) -> str:
        d = dict(knowledge)
        res = ''
        i = 0
        while i < len(s):
            if s[i] == '(':
                j = i + 1
                while j < len(s) and s[j] != ')':
                    j += 1
                key = s[i + 1:j]
                if key in d:
                    res += d[key]
                else:
                    res += '?'
                i = j
            else:
                res += s[i]
            i += 1
        return res
